fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/api/event/registry/DynamicRegistries.java

```diff
@@ -0,0 +1,154 @@
+import org.jetbrains.annotations.Unmodifiable;
+
+import net.minecraft.registry.Registry;
+import net.minecraft.registry.RegistryKey;
+import net.minecraft.registry.RegistryLoader;
+
+import net.fabricmc.fabric.impl.registry.sync.DynamicRegistriesImpl;
+
+/**
+ * Contai§ns methods for registering and accessing dynamic {@linkplain Registry registries}.
+ *
+ * <h2>Basic usage</h2>
+ * Custom dynamic registries can be registered with {@link #register(RegistryKey, Codec)}. These registries will not be
+ * <a href="#sync">synced to the client</a>.
+ *
+ * <p>The list of all dynamic registries, whether from vanilla or mods, can be accessed using
+ * {@link #getDynamicRegistries()}.
+ *
+ * <h2 id="sync">Synchronization</h2>
+ * Dynamic registries are not synchronized to the client by default.
+ * To register a <em>synced dynamic registry</em>, you can replace the {@link #register} call
+ * with a call to {@link #registerSynced(RegistryKey, Codec, SyncOption...)}.
+ *
+ * <p>If you want to use a different codec for syncing, e.g. to skip unnecessary data,
+ * you can use the overload with two codecs: {@link #registerSynced(RegistryKey, Codec, Codec, SyncOption...)}.
+ *
+ * <p>Synced dynamic registries can also be prevented from syncing if they have no entries.
+ * This is useful for compatibility with clients that might not have your dynamic registry.
+ * This behavior can be enabled by passing the {@link SyncOption#SKIP_WHEN_EMPTY} flag to {@code registerSynced}.
+ *
+ * <h2>Examples</h2>
+ * {@snippet :
+ * // @link region substring=RegistryKey target=RegistryKey
+ * // @link region substring=ofRegistry target="RegistryKey#ofRegistry"
+ * // @link region substring=Identifier target="net.minecraft.util.Identifier#Identifier(String, String)"
+ * public static final RegistryKey<Registry<MyData>> MY_DATA_KEY = RegistryKey.ofRegistry(new Identifier("my_mod", "my_data"));
+ * // @end @end @end
+ *
+ * // Option 1: Register a non-synced registry
+ * // @link substring=register target="#register":
+ * DynamicRegistries.register(MY_DATA_KEY, MyData.CODEC);
+ *
+ * // Option 2a: Register a synced registry
+ * // @link substring=registerSynced target="#registerSynced(RegistryKey, Codec, SyncOption...)":
+ * DynamicRegistries.registerSynced(MY_DATA_KEY, MyData.CODEC);
+ *
+ * // Option 2b: Register a synced registry with a different network codec
+ * // @link substring=registerSynced target="#registerSynced(RegistryKey, Codec, Codec, SyncOption...)":
+ * DynamicRegistries.registerSynced(MY_DATA_KEY, MyData.CODEC, MyData.NETWORK_CODEC);
+ * }
+ */
+public final class DynamicRegistries {
+	private DynamicRegistries() {
+	}
+
+	/**
+	 * Returns an unmodifiable list of all dynamic registries, including modded ones.
+	 *
+	 * <p>The list will not reflect any changes caused by later registrations.
+	 *
+	 * @return an unmodifiable list of all dynamic registries
+	 */
+	public static @Unmodifiable List<RegistryLoader.Entry<?>> getDynamicRegistries() {
```

Comment:
Annotation on a separate line?

Reply:
It's not an annotation on the method, it's an annotation on the type.

Reply:
Same as `@Nullable`. They're equivalent in bytecode I believe.

Reply:
No, it's different since `@Nullable` is also a method annotation. When you use `@Nullable` on a method return type, it makes it annotate both the return type and the method since it can be applied to either one. That's why it shows up twice for all usages in the Yarn javadoc, for example.

`@Unmodifiable` differs in that it's only a type use annotation, so it should go with the type IMO.

Reply:
Sure